### ISSUE TITLE
App won't start on version 2.4.8

### ISSUE THREAD

**Issue by ritox842** [angular/angular#14588]
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report 
[ ] feature request
[ ] support request
```

**Current behavior**

After update angular my application shows the loader and that's it, nothing happens. 
On version 2.4.7 the site works just fine.
I have no errors on console. I have no clue what's causing this issue except the version change
**Expected behavior**

The site should work just like it 2.4.7

**Minimal reproduction of the problem with instructions**

**What is the motivation / use case for changing the behavior?**

Updating angular from 2.4.7 to 2.4.8

**Please tell us about your environment:**

Windows 10, 
Webstorm 2016.3.3

* **Angular version:** 2.0.8

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Chrome 56.0.2924.87 (Only one that was tested) 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript 2.1.6

**Comment by DzmitryShylovich** [comment]
Pls add a repro

**Comment by ritox842** [comment]
A repro of what ? I can't link to the app cause it's a closed project..

**Comment by DzmitryShylovich** [comment]
what do you think we can do with it without the info how to reproduce that?  http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5?p=preview <- app using 2.4.8 and it works fine

**Comment by ASHFAQPATWARI** [comment]
I am also having the same issue. One of my nested route has a `canLoad guard` which is returning `false` initially and my app component is making a api request and based on that response  I redirect the user. It was working fine till 2.4.7 but now app.component is not initializing and hence this api request is not going.

**Comment by DzmitryShylovich** [comment]
Looks like it caused by https://github.com/angular/angular/pull/14327

**Comment by DzmitryShylovich** [comment]
but as I said it's very hard to do something without a small repro.
@ASHFAQPATWARI can you provide a plunkr?

**Comment by ASHFAQPATWARI** [comment]
can you point me to a plunker with route and route guards? Then I will do the necessary changes in it to reproduce this issue.

**Comment by volodymyro-in6k** [comment]
I have the same issue with  2.4.8.
It works only after canLoad or Resolve is turned off.

**Comment by ASHFAQPATWARI** [comment]
Yes, even if I return true in my canLoad, then things are working.

**Comment by DzmitryShylovich** [comment]
@ASHFAQPATWARI http://plnkr.co/edit/JWkl619QHbDiUtgec33t?p=preview

**Comment by ASHFAQPATWARI** [comment]
@DzmitryShylovich  http://plnkr.co/edit/RKMEWDN44rf8yoyYP2Ya?p=preview

There is a service called LoginService, in that make variable isLoggedIn true and code in app.component will initialize else not.

**Comment by fxck** [comment]
Not sure if it's related, but after updating to *.4.8, I'm getting errors caused by router events happening before root component's constructor is called (where I set up permission rules for my guards). This definitely was not happening on *.4.7.

**Comment by kirjai** [comment]
I too have stumbled upon this issue.. In our case, the following routing structure was causing the error:
```
const routes: Routes = [
  path: '',
  canActivate: [SomeGuard],
  children: [
      { path: '', redirect: 'some-route', pathMatch: 'full' },
      { path: 'some-route', loadChildren: '......' },
  ]
];
```
The issue was resolved as soon as I removed lazy loading of the redirect route.
```
...
{ path: 'some-route, component: SomeRouteComponent }
```

**Comment by saverett** [comment]
Experiencing a related issue. In 3.4.7 the `NavigationStart` event was emitted on the initial route. In 3.4.8 it no longer is.

**Comment by travisolbrich** [comment]
I am having similar issues. My auth guard returns an observable that doesn't emit until after I receive a successful token from an API call. Because of this the application fails to start and logs nothing to console.

**Comment by jameskentTX** [comment]
I'm writing a new application and have exactly the same problem. App works fine in 2.4.7 and nothing happens after upgrade to angular 2.4.8.
No visible errors, I see "Loading..." only (from index.html). I have no clue what is wrong in my case. I have returned to 2.4.7 version.

**Comment by Ristaaf** [comment]
We also have this problem, closed source app, but I can describe what we do and what the problem seems to be.

TL;DR

Our RouterModule.forChild routes have a resolve that depends on the constructor of the bootstrapped component to run before the promise is resolved. So the child routes will never be resolved before the application is bootstrapped, and with the changes in 3.4.8 version of the router the application will not be bootstrapped before the child routes is resolved, so for us the application stopped working with version 3.4.8, and even though we can probably change our implementation so that it starts working again, I would definitely call this a breaking change and it should have waited until version 4.  

What we do, If I have time I will create a plnkr for this and add it at the bottom:

We boostrap our AppComponent (Through an AppModule) which in it's constructor starts fetching all kinds of dependencies (setting up globalize with the correct locales, doing alot of System.imports etc.), when everything is fetched and ready we trigger an observable in a service called finishedLoadingDependencies.

The AppComponent has a router-outlet where we load our ShellComponent which in turn has it's own router-outlet where we load children using childRoutes, most of these children has a resolve part to their route definition, and this resolve part listens to the finishedLoadingDependencies observable mapped through a promise.

**Comment by AlexMokin** [comment]
Also have issue with updated router (3.4.8 version). We use custom UrlHandlingStrategy which has overridden shouldProcessUrl method which for some urls returns false. And when shouldProcessUrl  returns false application never gets bootstrapped.

**Comment by fulls1z3** [comment]
Having a more serious issue (with router 3.4.8 update) here:

Consider the following route configuration:

```
export const testRoutes: Routes = [
    {
        path: '',
        children: [
            {
                path: '',
                loadChildren: () => TestHomeModule
            },
            {
                path: 'about',
                loadChildren: () => TestAboutModule
            }
        ],
        data: {
            i18n: {
                isRoot: true
            }
        }
    },
    {
        path: 'change-language/:languageCode',
        component: TestComponent
    },
    {
        path: '**',
        redirectTo: '',
        pathMatch: 'full'
    }
];
```

The `''` and `'about'` routes are both lazy-loaded. If I navigate to `http://localhost:4200/` it's fine. I can also navigate to `http://localhost:4200/about` with a `routerLink`.

But if I type at the address bar `http://localhost:4200/about`, it bootstraps the app on `http://localhost:4200/` - instead it was `http://localhost:4200/about` on router 3.4.7.

This update simply ruined the location strategy.

**Comment by vsavkin** [comment]
Previously, the initial navigation would start after the root component has been created. Now it starts before it has been created. That's why if you have a CanLoad guard that is somehow initialized in the root component, it won't work. 

Unfortunately, this is the only way for us to wait for the navigation to complete on bootstrap, which is needed for universal.


These are your options:

### Do the initialization in the module

```
@NgModule({
  imports: [
    BrowserModule,
    RouterModule.forRoot(routes)
  ],
  declarations: [ App ],
  bootstrap: [ App ],
  providers: [ Guard, LoginService ]
})
export class AppModule {
  constructor(router: Router, loginService: LoginService) {
    loginService.initialize();
  }
}
```

This is a better strategy regardless of this issue. No reason to do in the component.



### Turns of the initial navigation and do it yourself

```
@Component({})
class App {
  constructor(router: Router, loginService: LoginService) {
    loginService.initialize();    
    router.initialNavigation();
  }
}

@NgModule({
  imports: [
    BrowserModule,
    RouterModule.forRoot(routes, {initialNavigation: false})
  ],
  declarations: [ App ],
  bootstrap: [ App ],
  providers: [ Guard, LoginService ]
})
export class AppModule {
}
```

**Comment by DzmitryShylovich** [comment]
> Unfortunately, this is the only way for us to wait for the navigation to complete on bootstrap, which is needed for universal.

is it possible to introduce some kind of a router bootstrap strategy so only universal users will be affected?

**Comment by vsavkin** [comment]
I think that the current way is actually the right way:
If my root component has an outlet, I would expect that outlet not to be empty. The empty outlet is a wrong state because there is no way for me to get into that state after the first navigation. The fact that we allowed it before was an accident. 

It's more important for universal users because it actually fails in that scenario.

**Comment by UnwrittenFun** [comment]
Can this change not be pushed back to v4? It is a breaking change as demonstrated by the number of people experiencing issues here, and also affected our production site.

**Comment by DzmitryShylovich** [comment]
yeah, it should be reverted in 2.x branch and pushed into 4.x with `breaking change` commit msg

**Comment by vsavkin** [comment]
@DzmitryShylovich We can introduce a separate flag to make the bootstrap logic behave as it does right now. So by default it will behave the same way it did in 3.4.7.

**Comment by DzmitryShylovich** [comment]
Flag is also a good option 👍

**Comment by kirjai** [comment]
I'm still unsure what the solution is for the lazy loaded routes. If the user tries to go to a lazy loaded route via the url bar, the `router-outlet` will always be empty, surely? I don't get how lazy loaded *can* work with this

**Comment by steveschmitt** [comment]
I tried using 
```
RouterModule.forRoot(routes, {initialNavigation: false})
```
in the `AppModule`, and 
```
router.initialNavigation();
```
in the `AppComponent`, as @vsavkin suggested.  But the AppComponent does not get instantiated at all if `initialNavigation` is false, so that doesn't work.  It's a Catch-22.

**Comment by steveschmitt** [comment]
IMHO Components that are outside of any `<router-outlet>` should not wait for the router to bootstrap.  They are logically separate from the router and should initialize independently.

**Comment by vsavkin** [comment]
@kirjai see here http://plnkr.co/edit/9z5cXGGXZ4eFSOlX1EuI?p=preview

@steveschmitt we do not resolve https://github.com/angular/angular/blob/2.4.x/modules/@angular/router/src/router_module.ts#L310. That's why my suggestion didn't work. It's a bug.

**Comment by steveschmitt** [comment]
@vsavkin [Here's an example](http://plnkr.co/edit/wBQq7C6bWuItqF4Cpz52?p=preview) of what our app is trying to do.  I've added a `LoginComponent` to your plunkr.  When the user clicks the button on the component, it calls `LoginService.initialize()` to set the `isLoggedIn` flag.  

Except that the user *can't* click the button, because the LoginComponent won't render, because the `canLoad` guard is blocking everything, even components that are not routed.
http://plnkr.co/edit/wBQq7C6bWuItqF4Cpz52?p=preview

**Comment by meeroslav** [comment]
@steveschmitt In your example there is no single route that is allowed for not logged in user. The fact that this is used to work for you was a side effect.
How can you show login button in the first place if you can't render the page that shows it (since guard keeps it away).

**Comment by bakasmarius** [comment]
I don't have any Guard but I can only see "Loading..." after update to 2.4.8.
I tried the solutions provided by @vsavkin but unsuccessfully: I get errors "Property 'initialNavigation' does not exist on type 'RouterModule'" and "Property 'initialize' does not exist on type 'LoginService'". LoginService is a service I created in my project and I know it doesn't have initialize function as it has only login and logout functions, but what about RouterModule?
Also, adding {initialNavigation: false} didn't seem to do anything.
What am I missing here?

**Comment by Ristaaf** [comment]
I solved my problem basically switching place between the classes in AppComponent and AppModule, Thanks Victor @vsavkin for that tip. 

However I still think this is a breaking change and should not have been made on a semver 3:PATCHFIX version, I guess you could argue that this is a bugfix and all of us just happened to rely on the functionality of a bug, however I would rather guess that you have pressure from another interest to include this and I totally sympathies with that.

**Comment by vicb** [comment]
@steveschmitt OTOH how do you expect it to work ? 

Let's imagine the Login and an empty outlet get rendered at first (I think this is what you expect ?) then what would happen when you click on log in (click on the button) ? How would the route be loaded ?

**Comment by vicb** [comment]
And just to make it clear, we understand that the change in behavior is causing troubles for a lot of user and we will most probably revert to the former behavior by default for the next 2.x release. Now you can keep your comments factual.

**Comment by steveschmitt** [comment]
@meeroslav if I had an unprotected route, could I configure the route to fallback to the unprotected route when the user does not have access to the requested route?

@vicb I would inject the router into the LoginComponent and use it to route after login.

**Comment by mogusbi** [comment]
Is there a timeline for when this fix will be released?

**Comment by JCornat** [comment]
I am experiencing this issue too, here a repro : http://plnkr.co/edit/bSvOlSjvGd8vH8RaEF9m?p=preview

**Comment by DzmitryShylovich** [comment]
@mogusbi the next release

**Comment by mogusbi** [comment]
@DzmitryShylovich OK let me rephrase my question then, when can we expect 2.4.9 with this fix to be released? :)

**Comment by juanmcampos** [comment]
@mogusbi based on https://github.com/angular/angular/blob/master/docs/RELEASE_SCHEDULE.md I think it would be next week :(, i should have never deleted my node_modules and updated the cli

**Comment by JCornat** [comment]
@juanmcampos You don't have to wait the next release, just fix your `package.json` dependencies to 2.4.7 instead of 2.4.8.

Example :

    "dependencies": {  
      "@angular/common": "2.4.7",  
      "@angular/compiler": "2.4.7",  
      "@angular/core": "2.4.7",  
      "@angular/forms": "2.4.7",  
      "@angular/http": "2.4.7",  
      "@angular/platform-browser": "2.4.7",  
      "@angular/platform-browser-dynamic": "2.4.7",  
      "@angular/platform-server": "2.4.7",  
      "@angular/router": "3.4.7",
      ...
    },
    "devDependencies": {
      "@angular/compiler-cli": "2.4.7",
      ...
    }`

**Comment by fulls1z3** [comment]
@DzmitryShylovich, @vsavkin I see that the sam set of changes causing this issue (41dd62c) have been included on 4.0.0-rc.1 (2a191ca):

```
router: do not finish bootstrap until all the routes are resolved (#14608) (2a191ca), closes #12162 #14155
```

Honestly we all had to fall back at 2.4.7 and it seems that won't be able to upgrade to a higher version of 4.0.0-beta.8 because these changes (2a191ca).

The question is, Angular 4.0 is planned to be backwards compatible, or not? Are there any plans about the changes on 2a191ca will exist, or be subject to opt-in, (or whatever) on further releases than rc.1?

Enlightment appreciated

**Comment by DzmitryShylovich** [comment]
@fulls1z3 https://github.com/angular/angular/issues/14588#issuecomment-281749508

> The question is, Angular 4.0 is planned to be backwards compatible, or not?

according to semver major releases may contain incompatible changes

**Comment by DzmitryShylovich** [comment]
@ritox842 the commit was reverted https://github.com/angular/angular/pull/14681
so you can close this issue

**Comment by ritox842** [comment]
Ok

**Comment by pravash-upreti** [comment]
@Component({})
class App {
  constructor(router: Router, loginService: LoginService) {
    loginService.initialize();    
    router.initialNavigation();
  }
}

@NgModule({
This is the best solution 👍 
Starting route after APP_Initialize response
  imports: [
    BrowserModule,
    RouterModule.forRoot(routes, {initialNavigation: false})
  ],
  declarations: [ App ],
  bootstrap: [ App ],
  providers: [ Guard, LoginService ]
})
export class AppModule {
}
